### ISSUE TITLE
AppPrivacyReport shouldn't be aware of network activities while using Leaf

### ISSUE THREAD

**Issue by edsburke** [eycorsican/leaf#302]
# Problem
App Privacy Report can record all network activities even if a Leaf based VPN is in use, however, this should not take place for the sake of privacy.

Note that, [App Privacy Report](https://support.apple.com/en-us/HT212958) is a new feature since iOS 15.2.

# How to Reproduce

1. On your iOS(v15.2+) device, install and connect [Orbot VPN](https://apps.apple.com/us/app/orbot/id1609461599) which is based on Leaf.
1. Ensure your internet facing IP has changed, as a proof Orbot is working.
1. Switch off App Privacy Report from iOS settings.
1. Switch on App Privacy Report.
1. Open Twitter and pick up a random post.
1. Check App Privacy Report,  in the section of APP NETWORK ACTIVITY, the all Twitter network access are recorded.

# Root Cause Analysis
This may be due to DNS exposure to iOS. 

Orbot uses TUN device supported by Leaf to route all traffic, chances are that DNS is resolved before traffic is sent to TUN, here is spot where DNS is exposed to iOS so it is aware of and records the apps' network activities. Here's the demo flow path for example:
```
DNS request --> iOS, DNS is exposed here
all other network request --> TUN
```
A counter proof is that all NEPacketTunnelFlow based VPN won't have this privacy issue, similarly.
```
all network request (including DNS) --> NEPacketTunnelFlow --> IP packet --> remote VPN server
```
# Proposal
If DNS exposure is the root cause, NEPacketTunnelFlow read/write interfaces are highly wanted and proposed. There's another ticket here https://github.com/eycorsican/leaf/issues/297

**Comment by eycorsican** [comment]
>  A counter proof is that all NEPacketTunnelFlow based VPN won't have this privacy issue, similarly.

I don't think so. `NEPacketTunnelFlow` is likely just another layer built upon a TUN device. So the traffic flow would look like the following if we use `NEPacketTunnelFlow`:

```
Apps -> TUN device -> NEPacketTunnelFlow -> Leaf
```

With the fd solution, however, it would look like:

```
Apps -> TUN device -> Leaf
```

Apple can collect your domain names in many ways. The most plausible way they can take is to collect from the local DNS resolver. Every time an app makes an https request, this DNS resolver is used. This is why they can associate the domains to a specific app.

Yet, iOS failed to collect domain names accessed by Leaf, because Leaf uses its own DNS resolver.

To make things clear, let's take the DNS resolution  into account and see how the traffic flow looks like:

```
┌─────┐                            ┌───────────────────┐
│     │     www.twitter.com        │                   │
│Apps ├───────────────────────────▲│Local DNS Resolver │
│     │▼───────────────────────────┤                   │
└──┬──┘     IP of www.twitter.com  └────────┬──────────┘
   │                                        │
   │                                        ▼
   │                               ┌──────────────────┐         ┌─────────────┐     ┌──────────┐
   │                               │                  │         │             │     │          │
   │       IP of www.twitter.com   │   TUN device     ├────────►│   Leaf      ├────►│Internet  │
   └───────────────────────────────►                  │         │             │     │          │
                                   └──────────────────┘         ├─────────────┤     └──────────┘
                                                                │ DNS Resolver│
                                                                └─────────────┘
```

**Comment by tladesignz** [comment]
Most apps use some high-level abstraction provided by iOS libraries like [`NSURLSession`](https://developer.apple.com/documentation/foundation/nsurlsession?language=objc) to connect to servers. This information could as well be grabbed in their implementations.

To be honest, I consider AppPrivacyReport a really nice feature to introduce some transparency on overreaching apps. I don't see, why we should block this in any way. It's off by default, anyway, so it's a very deliberate decision of the user.

**Comment by edsburke** [comment]
@tladesignz Thanks for this comment. Speaking of transparency on overreaching apps, AppPrivacyReport is a nice feature, however, it also incurs privacy concerns in the context of VPN, i.e. using [Orbot VPN](https://apps.apple.com/us/app/orbot/id1609461599) with AppPrivacyReport on, all website requests are recorded, which should not be because VPN is used for privacy. 

Although AppPrivacyReport is from the Apple, customers should not allow DNS leaking through any non-VPN apps.

Thoughts? Thanks.

**Comment by tladesignz** [comment]
> Although AppPrivacyReport is from the Apple, customers should not allow DNS leaking through any non-VPN apps.

Where does it leak DNS requests? The DNS requests also pass through the tunnel, AFAIK.

**Comment by edsburke** [comment]
> Where does it leak DNS requests? The DNS requests also pass through the tunnel, AFAIK.

When any non-VPN app is aware of your web requests such as DNS, it's a leaking strictly speaking. AppPrivacyReport is such an example in this case.

So back to this topic, my question is how we can implement a Leaf based VPN that makes AppPrivacyReport even not aware of any web requests in terms of DNS. The fact is [Orbot VPN](https://apps.apple.com/us/app/orbot/id1609461599) and [Leaf VPN](https://apps.apple.com/us/app/leaf-vpn/id1534109007) can't achieve that goal. 

Looking forward to your input @tladesignz

**Comment by tladesignz** [comment]
Sorry, but I find the word "leak" in this context quite irritating, not to say wrongly, and alarmistically used.

- Data is not leaving the device.
- Data is collected by the OS, through official means of the OS.
- Users have to switch this on and are fully aware of what they are doing. In other words, they *want* this behaviour.

There is nothing we can do here, because this is outside of the access of any VPN app (in other words, outside of the possibilities of Network Extensions). The OS controls the apps, not the other way around.

You're barking at the wrong tree.

We can reconsider, if you have any proof that data is actually leaving the device and that it is not by the intention of the user.

**Comment by eycorsican** [comment]
> The fact is [Orbot VPN](https://apps.apple.com/us/app/orbot/id1609461599) and [Leaf VPN](https://apps.apple.com/us/app/leaf-vpn/id1534109007) can't achieve that goal, some other NEPacketTunnelFlow based VPN can however.

I'm not going to say it's impossible to achieve this in an iOS VPN app. But it's improbable by simply switching to the NEPacketTunnelFlow. Could you point out which  VPN app is doing what you said?

**Comment by tladesignz** [comment]
Here, this is actually really bad, and we can't do anything about it:

https://twitter.com/mysk_co/status/1579997801047822336

**Comment by eycorsican** [comment]
> Here, this is actually really bad, and we can't do anything about it:
> 
> https://twitter.com/mysk_co/status/1579997801047822336

This is not new. I found this kind of issues a few years ago. They allow some routes which can allow IPs belonging to  WeChat to escape the VPN.

**Comment by eycorsican** [comment]
@tladesignz Found this blog post: https://blog.disconnect.me/ios-vpn-leak-advisory/

I guess WeChat was able to escape the VPN in my findings many years ago because it was using MPTCP, WeChat is an instant messaging app that would love to make use of MPTCP to maintain high availability of their service.

**Comment by eycorsican** [comment]
Since this issue has been inactive for a long time, it is now closed. Please reopen if necessary.
